File: web/src/api/vault.ts
import { apiRequest } from "./client";

export type MembershipTier = "free" | "vault_member" | "elite" | "founding_member";

export type ProductCollection = "soft_life" | "wealth" | "ceo" | "ai" | "inner_life" | "signature";

// Distinct from ProductCollection: this groups a member's *owned* products
// for the Dashboard/My Library nav, while ProductCollection organizes the
// public Shop catalog.
export type LifeArea = "soft_life" | "goals" | "money" | "ceo_life" | "ai" | "inner_life" | "challenges";

export interface Product {
  id: string;
  slug: string;
  title: string;
  subtitle: string;
  description: string;
  type: string;
  collection: ProductCollection;
  life_area: LifeArea;
  credit_line: string;
  price: number;
  thumbnail_url: string;
  is_ai_resource: boolean;
  is_monthly_drop: boolean;
  is_hero: boolean;
  drop_month: string | null;
}

export interface Bundle {
  id: string;
  slug: string;
  name: string;
  description: string;
  price: number;
  products: Product[];
  includes_app_access: boolean;
  is_founding_member: boolean;
  individual_total: number;
  savings: number;
}

export interface LibraryItem {
  product: Product;
  source: "product" | "bundle" | "subscription";
  granted_at: string;
  last_opened_at: string | null;
}

export interface Dashboard {
  welcome_message: string;
  membership_tier: MembershipTier;
  membership_badge: string | null;
  library_count: number;
  new_this_month: Product[];
  continue_your_journey: LibraryItem[];
}

export interface Drop {
  month: string;
  products: Product[];
  unlocked: boolean;
}

export interface Order {
  id: string;
  items: { type: "product" | "bundle"; title: string; price: number }[];
  amount: number;
  created_at: string;
}

export type BoxType = "skincare" | "lifestyle";
export type BoxTierKey = "mini" | "classic" | "deluxe";

export interface BoxTier {
  tier: BoxTierKey;
  label: string;
  price_low: number;
  price_high: number;
}

export interface Box {
  box_type: BoxType;
  label: string;
  teaser: string;
  tiers: BoxTier[];
  subscribed_tier: BoxTierKey | null;
}

export const vaultApi = {
  listProducts: (type?: string) =>
    apiRequest<Product[]>(`/vault/products${type ? `?type=${encodeURIComponent(type)}` : ""}`, { auth: false }),

  getProduct: (idOrSlug: string) => apiRequest<Product>(`/vault/products/${idOrSlug}`, { auth: false }),

  listBundles: () => apiRequest<Bundle[]>("/vault/bundles", { auth: false }),

  getBundle: (idOrSlug: string) => apiRequest<Bundle>(`/vault/bundles/${idOrSlug}`, { auth: false }),

  checkout: (payload: { product_id?: string; bundle_id?: string }) =>
    apiRequest<{ checkout_url: string; session_id: string }>("/vault/checkout", {
      method: "POST",
      body: payload,
    }),

  getDashboard: () => apiRequest<Dashboard>("/vault/dashboard"),

  getLibrary: (params?: { type?: string; lifeArea?: string }) => {
    const query = new URLSearchParams();
    if (params?.type) query.set("type", params.type);
    if (params?.lifeArea) query.set("life_area", params.lifeArea);
    const qs = query.toString();
    return apiRequest<LibraryItem[]>(`/vault/library${qs ? `?${qs}` : ""}`);
  },

  getDrops: () => apiRequest<Drop[]>("/vault/drops"),

  markOpened: (productId: string) =>
    apiRequest<void>(`/vault/products/${productId}/open`, { method: "POST" }),

  getDownloadUrl: (productId: string) =>
    apiRequest<{ download_url: string; expires_in_seconds: number }>(`/vault/products/${productId}/download`),

  getOrders: () => apiRequest<Order[]>("/vault/orders"),

  getBoxes: () => apiRequest<Box[]>("/vault/boxes"),

  checkoutBox: (payload: { box_type: BoxType; tier: BoxTierKey }) =>
    apiRequest<{ checkout_url: string; session_id: string }>("/vault/boxes/checkout", {
      method: "POST",
      body: payload,
    }),
};

File: backend/app/schemas/vault.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.utils.objectid import PyObjectId


class ProductOut(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    id: PyObjectId
    slug: str
    title: str
    subtitle: str
    description: str
    best_for: str
    outcome: str
    whats_inside: list[str]
    type: str
    collection: str
    life_area: str
    credit_line: str
    price: float
    thumbnail_url: str
    is_ai_resource: bool
    is_monthly_drop: bool
    is_hero: bool
    drop_month: str | None


class BundleOut(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    id: PyObjectId
    slug: str
    name: str
    description: str
    price: float
    products: list[ProductOut]
    includes_app_access: bool
    is_founding_member: bool
    individual_total: float
    savings: float


class LibraryItemOut(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    product: ProductOut
    source: str
    granted_at: datetime
    last_opened_at: datetime | None


class DashboardOut(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    welcome_message: str
    membership_tier: str
    membership_badge: str | None
    library_count: int
    new_this_month: list[ProductOut]
    continue_your_journey: list[LibraryItemOut]


class DropOut(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    month: str
    products: list[ProductOut]
    unlocked: bool


class OrderItemOut(BaseModel):
    type: str
    title: str
    price: float


class OrderOut(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    id: PyObjectId
    items: list[OrderItemOut]
    amount: float
    created_at: datetime


class CheckoutRequest(BaseModel):
    product_id: str | None = None
    bundle_id: str | None = None


class CheckoutResponse(BaseModel):
    checkout_url: str
    session_id: str


class CartItemIn(BaseModel):
    product_id: str | None = None
    bundle_id: str | None = None


class CartCheckoutRequest(BaseModel):
    items: list[CartItemIn] = Field(default_factory=list)


class DownloadResponse(BaseModel):
    download_url: str
    expires_in_seconds: int


class BoxTierOut(BaseModel):
    tier: str
    label: str
    price: int


class BoxOut(BaseModel):
    box_type: str
    label: str
    teaser: str
    tiers: list[BoxTierOut]
    subscribed_tier: str | None


class BoxCheckoutRequest(BaseModel):
    box_type: str
    tier: str


class BoxSubscriberOut(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    user_id: PyObjectId
    email: str
    full_name: str
    box_type: str
    tier: str
    created_at: datetime


class MonthPoint(BaseModel):
    month: str
    value: float


class GrowthOut(BaseModel):
    money: list[MonthPoint]
    goals: list[MonthPoint]
    wellness: list[MonthPoint]


class ReviewCreate(BaseModel):
    rating: int = Field(ge=1, le=5)
    title: str = ""
    body: str = ""
    display_name: str = ""


class ReviewOut(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    id: PyObjectId
    rating: int
    title: str
    body: str
    display_name: str
    verified_purchase: bool
    incentivized: bool
    created_at: datetime


class ProductReviewsOut(BaseModel):
    average_rating: float | None
    count: int
    reviews: list[ReviewOut]
